sentry_sdk/tracing.py

```diff
@@ -1,60 +1,136 @@
 import re
 import uuid
+import contextlib
+import collections
 
-_traceparent_header_format_re = re.compile(
-    "^[ \t]*([0-9a-f]{2})-([0-9a-f]{32})-([0-9a-f]{16})-([0-9a-f]{2})" "(-.*)?[ \t]*$"
-)
+from datetime import datetime
 
+from sentry_sdk.utils import concat_strings
 
-class _EnvironHeaders(object):
-    def __init__(self, environ):
-        self.environ = environ
 
-    def get(self, key):
-        return self.environ.get("HTTP_" + key.replace("-", "_").upper())
+if False:
+    from typing import Optional
+    from typing import Any
+    from typing import Dict
+    from typing import List
+
+_traceparent_header_format_re = re.compile(
+    "^[ \t]*"  # whitespace
+    "([0-9a-f]{32})?"  # trace_id
+    "-?([0-9a-f]{16})?"  # span_id
+    "-?([01])?"  # sampled
+    "[ \t]*$"  # whitespace
+)
 
 
-class SpanContext(object):
-    def __init__(self, trace_id, span_id, recorded=False, parent=None):
+class EnvironHeaders(collections.Mapping):
+    def __init__(self, environ, prefix="HTTP_"):
+        self.environ = environ
+        self.prefix = prefix
+
+    def __getitem__(self, key):
+        return self.environ[self.prefix + key.replace("-", "_").upper()]
+
+    def __len__(self):
+        return sum(1 for _ in iter(self))
+
+    def __iter__(self):
+        for k in self.environ:
+            if not isinstance(k, str):
+                continue
+
+            k = k.replace("-", "_").upper()
+            if not k.startswith(self.prefix):
+                continue
+
+            yield k[len(self.prefix) :]
+
+
+class Span(object):
+    __slots__ = (
+        "trace_id",
+        "span_id",
+        "parent_span_id",
+        "same_process_as_parent",
+        "sampled",
+        "transaction",
+        "op",
+        "description",
+        "start_timestamp",
+        "timestamp",
+        "_tags",
+        "_data",
+        "_finished_spans",
+    )
+
+    def __init__(
+        self,
+        trace_id,
+        span_id,
+        parent_span_id=None,
+        same_process_as_parent=True,
+        sampled=None,
+        transaction=None,
+        op=None,
```

Comment:
Shouldn't we make operation required?

Reply:
For transaction/trace spans, none of the Pytohn SDK's integrations set either OP or description. Only for child spans.

Reply:
If we want to make more attrs required I'd like to discuss what the difference between transaction and description is